File: Sources/MDifyCore/Services/WorkerClient.swift
import Foundation

public enum WorkerOCRMode: String, Equatable, Sendable {
    case auto
    case always
    case off
}

public protocol WorkerConverting: Sendable {
    func convert(inputURL: URL, outputURL: URL) async throws -> WorkerResponse
}

public struct WorkerClient: WorkerConverting {
    public let executableURL: URL
    public let kind: WorkerKind
    public let ocrMode: WorkerOCRMode
    public let runner: any ProcessRunning

    public init(
        executableURL: URL,
        kind: WorkerKind,
        ocrMode: WorkerOCRMode = .auto,
        runner: any ProcessRunning = ProcessRunner()
    ) {
        self.executableURL = executableURL
        self.kind = kind
        self.ocrMode = ocrMode
        self.runner = runner
    }

    public func convert(inputURL: URL, outputURL: URL) async throws -> WorkerResponse {
        var arguments = [
            "--input", inputURL.path,
            "--output", outputURL.path,
            "--format", "json"
        ]
        if kind == .ocr {
            arguments += ["--ocr", ocrMode.rawValue, "--ocr-lang", "cyrillic", "--dpi", "300"]
        }

        let result = try await runner.run(
            executableURL: executableURL,
            arguments: arguments,
            environment: nil
        )

        guard let data = result.stdout.data(using: .utf8), !data.isEmpty else {
            throw WorkerClientError.emptyResponse(result.stderr)
        }

        do {
            return try JSONDecoder().decode(WorkerResponse.self, from: data)
        } catch {
            throw WorkerClientError.invalidJSON(result.stdout + result.stderr)
        }
    }
}

public enum WorkerClientError: LocalizedError, Equatable {
    case emptyResponse(String)
    case invalidJSON(String)

    public var errorDescription: String? {
        switch self {
        case .emptyResponse(let stderr):
            stderr.isEmpty ? "Worker did not return JSON." : stderr
        case .invalidJSON:
            "Worker returned invalid JSON."
        }
    }
}

File: windows/MDify.Windows/Services/WorkerClient.cs
using System.Text.Json;
using MDify.Windows.Models;

namespace MDify.Windows.Services;

public enum WorkerOcrMode
{
    Auto,
    Always,
    Off
}

public interface IWorkerConverting
{
    WorkerKind WorkerKind { get; }

    Task<WorkerResponse> ConvertAsync(
        string inputPath,
        string outputPath,
        CancellationToken cancellationToken);
}

public sealed class WorkerClientException : Exception
{
    public WorkerClientException(string message, Exception? innerException = null)
        : base(message, innerException)
    {
    }
}

public sealed class WorkerClient : IWorkerConverting
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    private readonly string _executablePath;
    private readonly WorkerOcrMode _ocrMode;
    private readonly IProcessRunner _runner;

    public WorkerClient(
        string executablePath,
        WorkerKind workerKind,
        IProcessRunner? runner = null,
        WorkerOcrMode ocrMode = WorkerOcrMode.Auto)
    {
        _executablePath = executablePath;
        WorkerKind = workerKind;
        _runner = runner ?? new ProcessRunner();
        _ocrMode = ocrMode;
    }

    public WorkerKind WorkerKind { get; }

    public async Task<WorkerResponse> ConvertAsync(
        string inputPath,
        string outputPath,
        CancellationToken cancellationToken)
    {
        var arguments = new List<string>
        {
            "--input", inputPath,
            "--output", outputPath,
            "--format", "json"
        };

        if (WorkerKind == WorkerKind.Ocr)
        {
            arguments.AddRange(new[]
            {
                "--ocr", SerializeOcrMode(_ocrMode),
                "--ocr-lang", "cyrillic",
                "--dpi", "300"
            });
        }

        var result = await _runner.RunAsync(_executablePath, arguments, cancellationToken).ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(result.Stdout))
        {
            var message = string.IsNullOrWhiteSpace(result.Stderr)
                ? "Worker did not return JSON."
                : result.Stderr.Trim();
            throw new WorkerClientException(message);
        }

        try
        {
            return JsonSerializer.Deserialize<WorkerResponse>(result.Stdout, JsonOptions)
                ?? throw new WorkerClientException("Worker returned empty JSON.");
        }
        catch (JsonException error)
        {
            throw new WorkerClientException("Worker returned invalid JSON.", error);
        }
    }

    private static string SerializeOcrMode(WorkerOcrMode mode)
    {
        return mode switch
        {
            WorkerOcrMode.Auto => "auto",
            WorkerOcrMode.Always => "always",
            WorkerOcrMode.Off => "off",
            _ => throw new ArgumentOutOfRangeException(nameof(mode), mode, null)
        };
    }
}
